CIScripts/TimeLogger.ps1

```diff
@@ -0,0 +1,121 @@
+#
+# TimeLogger is a utility class that keeps track of time that parts of code
+# take to execute and prints them lazily.
+#
+# Class usage example:
+#
+# ```
+# $t = [TimeLogger]::new()
+# $t.Push("total")
+#     $t.Push("t0")
+#     $t.Pop()
+#     $t.Push("t1")
+#         $t.Push("t11")
+#         $t.Pop()
+#         $t.Push("t12")
+#         $t.Pop()
+#     $t.Pop()
+#     $t.Push("t2")
+#     $t.Pop()
+#     $t.Measure("ttt1", { echo "test123" })
+#     $t.LogAndMeasure("ttt2", { echo "test456" })
+# $t.Pop()
+# $t.Write()
+# ```
+#
+# console output:
+#
+# ```
+# test123
+# ttt2
+# test456
+# =======================================================
+#
+#  Time measurement results:
+#
+#           - 00:00:00.0099990: t0
+#                     - 00:00:00.0009985: t11
+#                     - 00:00:00: t12
+#           - 00:00:00.0049965: t1
+#           - 00:00:00: t2
+#           - 00:00:00.0229996: ttt1
+#           - 00:00:00.0070068: ttt2
+#   - 00:00:00.1469935: total
+#
+# =======================================================
+# ```
+
+class TimeMeasurement {
+    [string] $Name
+    [DateTime] $Start
+    [DateTime] $End
+
+    TimeMeasurement ([string] $Name) {
+        $this.Name = $Name
+    }
+
+    [TimeSpan] GetResult() {
+        return ($this.End - $this.Start)
+    }
+
+    [string] GetResultString([int] $IndentLevel) {
+        $msg = ""
+        1..($IndentLevel) | ForEach-Object{ $msg += " " }
+        $msg += "- " + ($this.GetResult()) + ": " + $this.Name + "`n"
+        return $msg
+    }
+}
+
+class TimeLogger {
+    [int] $CurrentIndentLevel
+    [int] $IndentJump
+    [string] $Output
+    [System.Collections.Stack] $Stack
+
+    TimeLogger () {
+        $this.Stack = New-Object System.Collections.Stack
+        $this.Stack.Push("asdf")
+        $this.Stack.GetType()
+        $this.CurrentIndentLevel = 0
+        $this.IndentJump = 10
+        $this.Output = ""
+    }
+
+    LogAndPush([string] $msg) {
+        Write-Host $msg
+        $this.Push($msg)
+    }
+
+    Push([string] $msg) {
+        $tm = [TimeMeasurement]::new($msg)
+        $tm.Start = Get-Date
+        $this.Stack.Push($tm)
+        $this.CurrentIndentLevel += $this.IndentJump
+    }
+
+    Pop() {
+        $tm = $this.Stack.Pop()
+        $tm.End = Get-Date
+        $this.CurrentIndentLevel -= $this.IndentJump
+        $this.Output += ($tm.GetResultString($this.CurrentIndentLevel))
```

Comment:
Would it be possible to reverse output printing? Currently it is presented like this:

```
<other results>
        - 00:03:19.5284725: Copying Agent dependencies
        - 00:00:20.2641599: Building API
        - 00:23:36.7044592: Building contrail-vrouter-agent.exe and .msi
        - 00:00:00.0310005: Building tests
- 00:27:16.5360921: Agent build
<other results>
```

But it makes more sense to print it like this:

```
<other results>
- 00:27:16.5360921: Agent build
        - 00:03:19.5284725: Copying Agent dependencies
        - 00:00:20.2641599: Building API
        - 00:23:36.7044592: Building contrail-vrouter-agent.exe and .msi
        - 00:00:00.0310005: Building tests
<other results>
```